backend/src/test/kotlin/org/loculus/backend/controller/submission/SubmitEndpointMultipartFileSharingTest.kt

```diff
@@ -0,0 +1,222 @@
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.loculus.backend.api.FileIdAndName
+import org.loculus.backend.config.BackendConfig
+import org.loculus.backend.config.BackendSpringProperty
+import org.loculus.backend.controller.DEFAULT_MULTIPART_FILE_PARTS
+import org.loculus.backend.controller.DEFAULT_ORGANISM
+import org.loculus.backend.controller.EndpointTest
+import org.loculus.backend.controller.S3_CONFIG
+import org.loculus.backend.controller.files.FilesClient
+import org.loculus.backend.controller.files.andGetFileIdsAndMultipartUrls
+import org.loculus.backend.controller.groupmanagement.GroupManagementControllerClient
+import org.loculus.backend.controller.groupmanagement.andGetGroupId
+import org.loculus.backend.controller.submission.SubmitFiles.DefaultFiles
+import org.loculus.backend.controller.submission.SubmitFiles.DefaultFiles.NUMBER_OF_SEQUENCES
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.http.MediaType.APPLICATION_JSON_VALUE
+import org.springframework.test.web.servlet.result.MockMvcResultMatchers.content
+import org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath
+import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
+import java.util.*
+
+@EndpointTest(
+    properties = ["${BackendSpringProperty.BACKEND_CONFIG_PATH}=$S3_CONFIG"],
+)
+class SubmitEndpointMultipartFileSharingTest(
+    @Autowired val submissionControllerClient: SubmissionControllerClient,
+    @Autowired val convenienceClient: SubmissionConvenienceClient,
+    @Autowired val filesClient: FilesClient,
+    @Autowired val backendConfig: BackendConfig,
+    @Autowired val groupManagementClient: GroupManagementControllerClient,
+) {
+    var groupId: Int = 0
+
+    @BeforeEach
+    fun prepareNewGroup() {
+        groupId = groupManagementClient.createNewGroup().andGetGroupId()
+    }
+
+    @Test
+    fun `GIVEN a valid multipart upload THEN the request is valid`() {
+        val fileIdAndUrls = filesClient.requestMultipartUploads(groupId, numberParts = 2)
+            .andGetFileIdsAndMultipartUrls()[0]
+        val etag1 = convenienceClient.uploadFile(fileIdAndUrls.presignedWriteUrls[0], DEFAULT_MULTIPART_FILE_PARTS[0])
+            .headers().map()["etag"]!![0]
+        val etag2 = convenienceClient.uploadFile(fileIdAndUrls.presignedWriteUrls[1], DEFAULT_MULTIPART_FILE_PARTS[1])
+            .headers().map()["etag"]!![0]
+
+        submissionControllerClient.submit(
+            DefaultFiles.metadataFile,
+            DefaultFiles.sequencesFile,
+            organism = DEFAULT_ORGANISM,
+            groupId = groupId,
+            fileMapping = mapOf(
+                "custom0" to mapOf(
+                    "myFileCategory" to listOf(
+                        FileIdAndName(
+                            fileIdAndUrls.fileId,
+                            "foo.txt",
+                            listOf(etag1, etag2),
```

Comment:
Commenting on the architecture not the test:

To me it would seem cleaner if the completion of a file was its own endpoint rather than happening here - then multipart and singlepart submissions would be the same it would just be the "uploading" that worked in a different way. I kind of feel when you batch upload a bunch of files you want to take each one of them through to "completion" as you upload it.

Reply:
(This isn't blocking, and is clearly a matter of taste, just having the discussion)

Reply:
I am happy with either solution. I'll ping the Slack to see whether others have preferences/arguments.

Reply:
I adapted the code so that there is now a `/files/complete-multipart-upload` endpoint. Thanks for the suggestion!